### PR TITLE
Create MultiSig GADT to accomodate for different eras

### PR DESCRIPTION
**Pull request by Jimbo4350**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Eras.hs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    -- used in many other places.
+  , Hash(PaymentKeyHash)
+
+  , PaymentKey
+  , PaymentExtendedKey
+  , GenesisKey
+  , GenesisUTxOKey
+  , GenesisDelegateKey
+  , StakeKey
+  , StakePoolKey
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Eras.hs:28]: I know this is a draft, but these exports don't seem all that relevant to an `Eras` module. They seem like they'd fit better as part of a `Key` module or something.

Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Eras.hs:28]: Agree, this is temporary though. I will likely do as you suggest and break off a `Key` module or something similar.

File: cardano-api/src/Cardano/API.hs [hunk 2]
```diff
@@ -135,12 +141,18 @@ module Cardano.API (
 
     -- ** Script addresses
     -- | Making addresses from scripts.
-    scriptHash,
+    scriptHashShelley,
+    scriptHashAllegra,
+    scriptHashMary,
 
-    -- ** Multi-signature scripts
+    -- ** Multi signature scripts
```
Comment (intricate) [inline cardano-api/src/Cardano/API.hs:148]: Nitpick, but:

```suggestion
    -- ** Multi-signature scripts
```

was probably appropriate.

**Comment by dcoutts** [comment]
I think the first patch in this PR is perfect and we should merge it now, while we continue to think about the API details for scripts.

**Comment by Jimbo4350** [comment]
Subsumed by #2069

**Pull request closed without merging**